RecoLocalCalo/HcalRecAlgos/src/MahiGPU.cu

```diff
@@ -0,0 +1,1857 @@
+        ? did2linearIndexHB(id, maxDepthHB, firstHBRing, lastHBRing, nEtaHB)
+        : did2linearIndexHE(id, maxDepthHE, maxPhiHE, firstHERing, lastHERing, nEtaHE)
+            + offsetForHashes;
+
+    auto const* pedestalWidthsForChannel = useEffectivePedestals && gch < nchannelsf01HE
+        ? effectivePedestalWidths + hashedId*4
+        : pedestalWidths + hashedId*4;
+    auto const averagePedestalWidth2 = 0.25 * (
+        pedestalWidthsForChannel[0]*pedestalWidthsForChannel[0] +
+        pedestalWidthsForChannel[1]*pedestalWidthsForChannel[1] +
+        pedestalWidthsForChannel[2]*pedestalWidthsForChannel[2] +
+        pedestalWidthsForChannel[3]*pedestalWidthsForChannel[3]);
+    auto const* gains = gainValues + 
+        hashedId*4;
+    // FIXME on cpu ts 0 capid was used - does it make any difference
+    auto const gain = gains[0];
+    auto const respCorrection = respCorrectionValues[hashedId];
+
+#ifdef HCAL_MAHI_GPUDEBUG
+#ifdef HCAL_MAHI_GPUDEBUG_FILTERDETID
+    if (id != DETID_TO_DEBUG) return;
+#endif
+#endif
+    
+    // TODO: provide this properly
+    int const soi = soiSamples[gch];
+    constexpr float deltaChi2Threashold = 1e-3;
+
+    ColumnVector<NPULSES, int> pulseOffsets;
+    #pragma unroll
+    for (int i=0; i<NPULSES; ++i)
+        pulseOffsets(i) = i;
+//        pulseOffsets(i) = pulseOffsetValues[i] - pulseOffsetValues[0];
+
+    // output amplitudes/weights
+    ColumnVector<NPULSES> resultAmplitudesVector = ColumnVector<NPULSES>::Zero();
+
+    // map views
+    Eigen::Map<const ColumnVector<NSAMPLES>> 
+        inputAmplitudesView{inputAmplitudes + gch*NSAMPLES};
+    Eigen::Map<const ColumnVector<NSAMPLES>> 
+        noiseTermsView{noiseTerms + gch*NSAMPLES};
+    Eigen::Map<ColMajorMatrix<NSAMPLES, NPULSES>> 
+        pulseMatrixMView{pulseMatricesM + gch*NSAMPLES*NPULSES};
+    Eigen::Map<ColMajorMatrix<NSAMPLES, NPULSES>>
+        pulseMatrixPView{pulseMatricesP + gch*NSAMPLES*NPULSES};
+    Eigen::Map<ColMajorMatrix<NSAMPLES, NPULSES>> 
+        pulseMatrixView{pulseMatrices + gch*NSAMPLES*NPULSES};
+
+#ifdef HCAL_MAHI_GPUDEBUG
+    for (int i=0; i<NSAMPLES; i++)
+        printf("inputValues(%d) = %f noiseTerms(%d) = %f\n", 
+            i, inputAmplitudesView(i), i, noiseTermsView(i));
+    for (int i=0; i<NSAMPLES; i++) {
+        for (int j=0; j<NPULSES; j++)
+            printf("%f ", pulseMatrixView(i, j));
+        printf("\n");
+    }
+    printf("\n");
+    for (int i=0; i<NSAMPLES; i++) {
+        for (int j=0; j<NPULSES; j++)
+            printf("%f ", pulseMatrixMView(i, j));
+        printf("\n");
+    }
+    printf("\n");
+    for (int i=0; i<NSAMPLES; i++) {
+        for (int j=0; j<NPULSES; j++)
+            printf("%f ", pulseMatrixPView(i, j));
+        printf("\n");
+    }
+#endif
+
+    int npassive = 0;
+    float chi2=0, previous_chi2=0.f, chi2_2itersback=0.f;
+    // TOOD: provide constants from configuration
+    for (int iter=1; iter<50; iter++) {
```

Comment:
as a general note in the CPU version this is 500 and not 50

see RecoLocalCalo/HcalRecProducers/python/HBHEMahiParameters_cfi.py
    nMaxItersMin      = cms.int32(500)